gap/SnAnUnknownDegree.gi

```diff
@@ -0,0 +1,710 @@
+# FIXME: Move me into GAP
+# Helper function to compute all primes up to a given integer via a prime sieve
+BindGlobal("AllPrimesUpTo",
+function(n)
+    local i, j, sieve, result;
+    if n <= 1000 then
+      return Primes{[1..PositionSorted(Primes, n+1)-1]};
+    fi;
+    sieve := BlistList([1..n], [1..n]);
+    sieve[1] := false;
+    for i in [2..Int(n/2)] do
+        sieve[i*2] := false;
+    od;
+    i := 3;
+    while i * i <= n do
+        if sieve[i] then
+            j := 3*i;
+            while j <= n do
+                sieve[j] := false;
+                j := j + 2*i;
+            od;
+        fi;
+        i := i + 2;
+    od;
+    return ListBlist([1..n], sieve);
+end);
+
+
+# ri : recog info record with group G
+# eps : real number, the error bound
+# N : integer, upper bound for the degree of G
+#
+# The following algorithm constructs a set of possible 3-cycles. It is based
+# on the simple observation that the product of two involutions t1, t2, which
+# only move one common point, squares to a 3-cycle.
+#
+# TODO: take care of duplicate candidates?
+# Creates and returns a function, here called oneThreeCycleCandidate. The
+# function oneThreeCycleCandidate returns one of the following:
+# - a three cycle candidate, i.e. an element of G
+# - TemporaryFailure, if we exhausted all attempts
+# - NeverApplicable, if we found out that G can't be an Sn or An
+BindGlobal("ThreeCycleCandidatesIterator",
+    function(ri, eps, N)
+    local
+        # involution
+        t,
+        # integers, controlling the number of iterations
+        M, B, T, C, logInt2N,
+        # iteration over all primes up to N
+        allPrimes, i, p,
+        # counters
+        nrInvolutions, nrTriedConjugates, nrThreeCycleCandidates,
+        # helper functions
+        tryThreeCycleCandidate, oneThreeCycleCandidate;
+    # Step 1: Initialization
+    # The current involution t_i
+    t := fail;
+
+    # Constants
+    # TODO: better iteration over primes
+    M := 1;
+    allPrimes := AllPrimesUpTo(N);
+    for i in [2 .. Length(allPrimes)] do
+        p := allPrimes[i];
+        M := M * p ^ LogInt(N, p);
+    od;
```

Comment:
The value `M` you computer here only depends on the positive integer `N`, right? And you recompute it here again and again. This suggests to me that perhaps using `MemoizePosIntFunction` here might be beneficial, esp. since `ThreeCycleCandidatesIterator` is likely to be called repeatedly. At the same time, N likely isn't going to be very big, is it? (like, often  a few ten thousand perhaps?)

Actually, that reminds me: we should also think about how the heuristic for coming up with an initial value for N could like? For permutation groups, of course it's easy (`NrMovedPoints(G)`). What does one do for matrix groups over a finite field? A naive approach is this: for a subgroup of GL(d,q),  we must have that Factorial(N) divides the size of that GL(d,q), which we know, so we can compute the largest N with this property. But I assume one can do much better (e.g. using representation theory,  knowledge about element orders, ...).